src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -66,120 +80,118 @@ func DeserializeState(x Bytes) State {
-		LockedPledgeCollateral_: actor.TokenAmount(0),
-	}
-
-	// TODO: call constructor of StorageMinerActor
-	// store ownerAddr and workerAddr there
-	// and return StorageMinerActor address
+	rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
 
-	minerID := rt.ImmediateCaller()
-	panic("TODO: caller is not miner actor")
+	// ownerAddr := rt.ImmediateCaller()
+	msgValue := rt.ValueReceived()
 
-	h, st := a.State(rt)
+	var newMinerAddr addr.Address
+	TODO() // TODO: call InitActor::Exec to create the StorageMinerActor
+	panic("")
+	// TODO: also store libp2p.PeerID here? (or should this be part of the miner actor's own state?)
 
-	st.PowerTable()[minerID] = newMiner
+	// TODO: anything to check here?
+	newMinerEntry := &PowerTableEntry_I{
+		ActivePower_:   block.StoragePower(0),
+		InactivePower_: block.StoragePower(0),
+	}
 
+	h, st := a.State(rt)
+	newTable, ok := actor.BalanceTable_WithNewAddressEntry(st.EscrowTable(), newMinerAddr, msgValue)
+	if !ok {
+		panic("Internal error: newMinerAddr (result of InitActor::Exec) already exists in escrow table")
+	}
+	st.Impl().EscrowTable_ = newTable
+	st.PowerTable()[newMinerAddr] = newMinerEntry
 	UpdateRelease(rt, h, st)
 
-	return minerID
-
+	return newMinerAddr
 }
 
-func (a *StoragePowerActorCode_I) RemoveStorageMiner(rt Runtime, address addr.Address) {
+func (a *StoragePowerActorCode_I) RemoveStorageMiner(rt Runtime) {
+	rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_StorageMiner)
 
-	minerID := rt.ImmediateCaller()
-	panic("TODO: use address and verify address is the caller")
-	panic(minerID)
+	minerAddr := rt.ImmediateCaller()
 
 	h, st := a.State(rt)
 
-	if (st.PowerTable()[address].ActivePower() + st.PowerTable()[address].InactivePower()) > 0 {
-		rt.AbortStateMsg("power still remains.")
+	minerPowerTotal, ok := st._getPowerTotalForMiner(minerAddr)
+	if !ok {
+		rt.AbortArgMsg("spa.RemoveStorageMiner: miner entry not found")
 	}
+	if minerPowerTotal > 0 {
+		// TODO: manually remove the power entries here (and update relevant counters),
+		// instead of throwing a runtime error?
+		rt.AbortStateMsg("power still remains.")
 
-	delete(st.PowerTable(), address)
-
-	UpdateRelease(rt, h, st)
-}
-
-func (a *StoragePowerActorCode_I) GetTotalPower(rt Runtime) block.StoragePower {
-
-	totalPower := block.StoragePower(0)
+		TODO()
+		// TODO: also fail if funds still remaining in escrow?
+	}
 
-	h, st := a.State(rt)
+	delete(st.PowerTable(), minerAddr)
 
-	for _, miner := range st.PowerTable() {
-		totalPower = totalPower + miner.ActivePower() + miner.InactivePower()
+	newTable, ok := actor.BalanceTable_WithDeletedAddressEntry(st.EscrowTable(), minerAddr)
+	if !ok {
+		panic("Internal error: miner entry in escrow table does not exist")
 	}
+	st.Impl().EscrowTable_ = newTable
 
-	Release(rt, h, st)
-
-	return totalPower
+	UpdateRelease(rt, h, st)
 }
 
 func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) {
```

Comment:
`st._ensurePledgeCollateralSatisfied` is removed from this PR so we need to update the function here too. `EnsurePledgeCollateral` is called during `SubmitVerfiedSurprisePoSt` to ensure that miner has sufficient pledge collateral at that point. Otherwise, the PoSt may fail and miner needs to `AddBalance` before submitting PoSt again.